File: Sprint5/I/C++/main.cpp
// https://contest.yandex.ru/contest/24809/problems/I/
#include <iostream>
#include <vector>
using namespace std;

long long get_num_bin_search_trees(int n) {
    vector<long long> nums(n + 1, 0);
    nums[0] = nums[1] = 1;

    for (int i = 2; i <= n; i++) {
        for (int j = 0; j < i; j++) {
            nums[i] += nums[j] * nums[i - j - 1];
        }
    }

    return nums[n];
}

int main() {
    int n;
    std::cin >> n;
    std::cout << get_num_bin_search_trees(n) << std::endl;

    return 0;
}

File: Sprint5/I/Rust/src/main.rs
// https://contest.yandex.ru/contest/24809/problems/I/

fn get_num_bin_search_trees(n: usize) -> u64 {
    let mut nums = vec![0; n + 1];
    nums[0] = 1u64;
    nums[1] = 1u64;

    for i in 2..n + 1 {
        for j in 0..i {
            nums[i] += nums[j] * nums[i - j - 1];
        }
    }

    nums[n]
}

fn main() {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let n = input.trim().parse::<usize>().unwrap();

    println!("{}", get_num_bin_search_trees(n));
}
